### PR TITLE
Remove `--inspect` & `--inspect-brk` from `execArgv`

### PR DESCRIPTION
**Pull request by wardpeet**
When using `execa.node` and the parent process has --inspect or --inspect-brk defined, it will run the child process with those options as well. It will throw an error that the inspector already is running on port 9229.

I've added a filter for those options on the default execArgv value. I tried to test it but I'm not sure this is the correct way to do so.

### CODE REVIEW

File: test/node.js [hunk 1]
```diff
@@ -3,7 +3,8 @@ import test from 'ava';
 import pEvent from 'p-event';
 import execa from '..';
 
-process.env.PATH = path.join(__dirname, 'fixtures') + path.delimiter + process.env.PATH;
+process.env.PATH =
+	path.join(__dirname, 'fixtures') + path.delimiter + process.env.PATH;
```
Comment (sindresorhus) [inline test/node.js:7]: Don't do unrelated changes.

File: test/node.js [hunk 2]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
...
+	async t => {
+		const {stdout, stderr} = await execa.node('console.log("foo")', {
+			stdout: 'pipe',
+			nodeOptions: ['--inspect', '-e']
+		});
+
+		t.is(stdout, 'foo');
+		t.regex(
+			stderr,
+			/^Debugger listening on ws:\/\/127.0.0.1:9229\/.*[\r\n]+For help, see: https:\/\/nodejs.org\/en\/docs\/inspector$/
```
Comment (sindresorhus) [inline test/node.js:127]:
```suggestion
			/Debugger listening/
```

I don't want to have to change it if it slightly changes on the Node.js side.

File: test/node.js [hunk 3]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
...
+		t.is(stderr, '');
+	}
+);
+
+test.serial(
+	'node removes --inspect-brk=9222 from nodeOptions when defined by parent process',
+	async t => {
+		const originalArgv = process.execArgv;
+		process.execArgv = ['--inspect-brk=9222', '-e'];
+		const childProc = execa.node('console.log("foo")', {
```
Comment (sindresorhus) [inline test/node.js:100]:
```suggestion
		const subprocess = execa.node('console.log("foo")', {
```

File: index.js [hunk 4]
```diff
@@ -235,7 +235,24 @@ module.exports.node = (scriptPath, args, options = {}) => {
...
+			return false;
+		}
+
+		return true;
+	});
+
+	const {
+		nodePath = process.execPath,
+		nodeOptions = defaultExecArgv
+	} = options;
```
Comment (sindresorhus) [inline index.js:255]: Any thoughts on doing it no matter what? I don't really see any good reason users would want to override it.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js [hunk 5]
```diff
@@ -235,7 +235,24 @@ module.exports.node = (scriptPath, args, options = {}) => {
 
 	const stdio = normalizeStdio.node(options);
 
-	const {nodePath = process.execPath, nodeOptions = process.execArgv} = options;
+	// Filter out `--inspect` & `--inspect-brk` from default `execArgv`.
+	const defaultExecArgv = process.execArgv.filter(arg => {
```
Comment (ehmicky) [inline index.js:239]: The following might be simpler:

```js
const defaultExecArgv = process.execArgv.filter(arg => arg !== '--inspect' && !arg.startsWith('--inspect=') && arg !== '--inspect-brk' && !arg.startsWith('--inspect-brk='))
```

File: index.js [hunk 6]
```diff
@@ -235,7 +235,24 @@ module.exports.node = (scriptPath, args, options = {}) => {
...
+		if (
+			arg === '--inspect' ||
```
Comment (ehmicky) [inline index.js:241]: I think using only `arg.startsWith('--inspect')` would work here since no `process.execArgv` only targets Node CLI flags, and only `--inspect` and `--inspect-brk` start with `--inspect`. If new `--inspect-*` are added in the future, we most likely will want to exclude them as well.

File: index.js [hunk 7]
```diff
@@ -235,7 +235,24 @@ module.exports.node = (scriptPath, args, options = {}) => {
 
 	const stdio = normalizeStdio.node(options);
 
-	const {nodePath = process.execPath, nodeOptions = process.execArgv} = options;
+	// Filter out `--inspect` & `--inspect-brk` from default `execArgv`.
```
Comment (ehmicky) [inline index.js:238]: I think this comment might not be necessary since the code is quite clear regardless.

File: test/node.js [hunk 8]
```diff
@@ -19,11 +20,14 @@ test('node pipe stdout', async t => {
...
-		nodeOptions: process.platform === 'win32' ? ['/c'] : []
-	});
+	const {stdout} = await execa.node(
+		process.platform === 'win32' ? 'hello.cmd' : 'hello.sh',
+		{
+			stdout: 'pipe',
+			nodePath: process.platform === 'win32' ? 'cmd.exe' : 'bash',
+			nodeOptions: process.platform === 'win32' ? ['/c'] : []
+		}
+	);
```
Comment (ehmicky) [inline test/node.js:30]: Same comment as https://github.com/sindresorhus/execa/pull/435/files#r496222326

File: test/node.js [hunk 9]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
 	t.is(stdout, 'foo');
 });
 
+test.serial(
```
Comment (ehmicky) [inline test/node.js:44]: Those tests do not need to be `test.serial()`, unless I am mistaken?

File: test/node.js [hunk 10]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
...
+	'node removes --inspect from nodeOptions when defined by parent process',
+	async t => {
+		const originalArgv = process.execArgv;
+		process.execArgv = ['--inspect', '-e'];
+		const {stdout, stderr} = await execa.node('console.log("foo")', {
+			stdout: 'pipe'
+		});
+		process.execArgv = originalArgv;
```
Comment (ehmicky) [inline test/node.js:52]: This statement should be inside `try`/`finally` in case `execa.node()` throws (due to a bug for example) so it does not impact the other tests.

File: test/node.js [hunk 11]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
 	t.is(stdout, 'foo');
 });
 
+test.serial(
+	'node removes --inspect from nodeOptions when defined by parent process',
+	async t => {
+		const originalArgv = process.execArgv;
+		process.execArgv = ['--inspect', '-e'];
+		const {stdout, stderr} = await execa.node('console.log("foo")', {
+			stdout: 'pipe'
```
Comment (ehmicky) [inline test/node.js:50]: `stdout` defaults to `pipe`, so does not need to be defined here (and in the following tests).

File: test/node.js [hunk 12]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
...
+
+test.serial(
+	'node removes --inspect-brk from nodeOptions when defined by parent process',
+	async t => {
+		const originalArgv = process.execArgv;
+		process.execArgv = ['--inspect-brk', '-e'];
+		const childProc = execa.node('console.log("foo")', {
+			stdout: 'pipe'
+		});
+		process.execArgv = originalArgv;
```
Comment (ehmicky) [inline test/node.js:82]: The few lines above are repeated across tests: would it be possible to create an helper function instead?

File: test/node.js [hunk 13]
```diff
@@ -37,6 +41,94 @@ test('node pass on nodeOptions', async t => {
...
+
+		setTimeout(() => {
```
Comment (ehmicky) [inline test/node.js:84]: `async`/`await` could be used here:

```js
const { setTimeout } = require('util');
const pSetTimeout = promisify(setTimeout);
```

Then:

```js
const subprocess = execa.node(..., { reject: false });
```

Then:

```js
await pSetTimeout(1000);
subprocess.kill();
const {stdout, stderr} = await subprocess
```

File: test/node.js (hunk 2 above)
Comment (ehmicky) [reply test/node.js:127]: This would also allow using `string.includes()` directly, which might be simpler:

```js
t.true(stderr.includes('Debugger listening'));
```

File: test/node.js (hunk 1 above)
Comment (wardpeet) [reply test/node.js:7]: it's my vscode :'( sorry i'll revert

File: index.js (hunk 5 above)
Comment (wardpeet) [reply index.js:239]: I find this harder to read but happy to update it.

File: index.js (hunk 4 above)
Comment (wardpeet) [reply index.js:255]: Sometimes you want to set inspect on a subprocess when your main process is just a small wrapper. We could also tell people to move towards https://nodejs.org/api/inspector.html instead for this functionality.

I'm okay either way, just remember that changing this would be a breaking change.

Comment (ehmicky) [reply index.js:255]: > Sometimes you want to set inspect on a subprocess when your main process is just a small wrapper. 

Yes, it looks to me that could be the case, and this change would then be a breaking change if any users is doing this. Based on this, I personally think only doing this as a default value (what the PR currently does) might be better?

File: test/node.js [hunk 14]
```diff
@@ -37,6 +44,75 @@ test('node pass on nodeOptions', async t => {
...
+
+test.serial(
+	'node removes --inspect-brk=9222 from nodeOptions when defined by parent process',
+	async t => {
+		process.execArgv = ['--inspect-brk=9222', '-e'];
+		const subprocess = execa.node('console.log("foo")', {
+			reject: false
+		});
+
+		const {stdout, stderr} = await subprocess.catch(error => error);
```
Comment (ehmicky) [inline test/node.js:96]: Is `.catch()` needed considering the `reject: false` option is used?

File: test/node.js [hunk 15]
```diff
@@ -37,6 +44,75 @@ test('node pass on nodeOptions', async t => {
...
+	async t => {
+		process.execArgv = ['--inspect', '-e'];
+		const {stdout, stderr} = await execa.node('console.log("foo")', {
+			reject: false
+		});
+
+		t.is(stdout, 'foo');
+		t.is(stderr, '');
+	}
+);
```
Comment (ehmicky) [inline test/node.js:58]: The tests (except the last one) are almost identical except for the value of `process.execArgv`.
Using [test macros](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#reusing-test-logic-through-macros) might reduce code size here.

File: test/node.js [hunk 16]
```diff
@@ -5,6 +5,13 @@ import execa from '..';
 
 process.env.PATH = path.join(__dirname, 'fixtures') + path.delimiter + process.env.PATH;
 
+test.beforeEach(t => {
+	t.context.originalArgv = process.execArgv;
+});
+
+test.afterEach(t => {
+	process.execArgv = t.context.originalArgv;
+});
```
Comment (ehmicky) [inline test/node.js:14]: If using test macros (see comment below), a `try`/`finally` block inside the tests could be used instead of `beforeEach`/`afterEach` hooks.

**Review by ehmicky — APPROVED** [review]
This looks good to me :+1: 
What do you think @sindresorhus ?

**Comment by wardpeet** [comment]
Thanks! I wasn't familiar with ava so sorry for taking so long 👍

**Comment by sindresorhus** [comment]
Nice work, @wardpeet
